sos/collector/transports/saltstack.py

```diff
@@ -0,0 +1,140 @@
+# Copyright Red Hat 2022, Trevor Benson <trevor.benson@gmail.com>
+
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+import contextlib
+import json
+import os
+from sos.collector.transports import RemoteTransport
+from sos.collector.exceptions import ConnectionException
+from sos.utilities import (is_executable,
+                           sos_get_command_output)
+
+TIMEOUT_DEFAULT = 300
+
+
+class SaltStackMasterTransport(RemoteTransport):
+    """
+    A transport for collect that leverages SaltStack's Master Pub/Sub
+    functionality to send commands to minions.
+
+    This transport will by default assume the use cmd.shell module to
+    execute commands on the minions.
+    """
+
+    name = 'saltstack'
+
+    def _convert_output_json(self, json_output):
+        return list(json.loads(json_output).values())[0]
+
+    def run_command(
+            self, cmd, timeout=180, need_root=False, env=None, get_pty=False):
+        """
+        Run a command on the remote host using SaltStack Master.
+        If the output is json, convert it to a string.
+        """
+        ret = super(SaltStackMasterTransport, self).run_command(
+            cmd, timeout, need_root, env, get_pty)
+        with contextlib.suppress(Exception):
+            ret['output'] = self._convert_output_json(ret['output'])
+        return ret
+
+    def _salt_retrieve_file(
+            self, node, fname, dest,
+            timeout=180, need_root=False, env=None, get_pty=False):
+        """
+        Execute cp.push on the remote host using SaltStack Master
+        """
+        cmd = f"salt {node} cp.push {fname}"
+        res1 = sos_get_command_output(cmd)
+        if res1['status'] == 0:
+            cachedir = f"/var/cache/salt/master/minions/{self.address}/files"
+            cachedir_file = os.path.join(cachedir, fname.lstrip('/'))
+            cmd = f"mv {cachedir_file} {dest}"
+            res2 = sos_get_command_output(cmd)
+        return res2['status'] == 0
```

Comment:
This needs to be indented into the `if res1['status'] == 0` conditional. If `res1['status']` is not `0`, then we'll hit a `NameError` as `res2` is yet to be defined.

If `res1['status']` is not `0`, then we should explicitly return False. I think `shutil` here is better than directly calling `mv`, also.

Something like this (untested):

```python
cmd = f"salt {node} cp.push {fname}"
res = sos_get_command_output(cmd)
if res['status'] == 0:
    cachedir = f"/var/cache/salt/master/minions/{self.address}/files"
    cachedir_file = os.path.join(cachedir, fname.lstrip('/'))
    shutil.move(cachedir_file, dest)
    return True
return False
```

Reply:
I'll nest it. Does this comment also mean to suggest usage of **shutil.move()** instead of **sos_get_command_output()** which replaced shutil?

Reply:
Doh, reading the comment in vscode, missed the explicit statement about shutil, fixing.